### ISSUE TITLE
We need a maintenance branch for 1.8

### ISSUE THREAD

**Issue by medoc92** [mrjimenez/pupnp#93]
The 1.8 version of the library is set to become the main version in the near future, becoming the default package on some Linux distributions.

It is time that it goes into maintenance mode on a branch which will only receive bug and security fixes.

Other changes should go to the master branch (future 1.9 or 1.10).

Some of the recent changes to 1.8 have been quite infortunate for a production branch (API ! and significant behaviour changes).

**Comment by mrjimenez** [comment]
The purpose of 1.8 was to adopt a purely functional API with encapsulated data such that there should never again be necessary to break API, which used to be something very common in the old days. Unfortunately, we had a problem with the callbacks which I hope will not happen again.

As for the significant behavior changes, maybe you could be more specific about what has displeased you.

I have no hard feelings about creating a new branch, but I don't really feel the need to. The only new feature that has appeared was the "ExtraHeaders" branch (philippe44-master), which has been stopped for a while, and since no one tried to push it forward it remains in its own branch. When it is ready, maybe we could discuss a new line of development, or maybe include it in 1.8 since it does not seem to break API.

Anyway, as always, I am open to other opinions.

Regards,
Marcelo.

**Comment by medoc92** [comment]
With the addition of the compatibility macros, it has become easy to switch from 1.6 to 1.8, with the exception of the vdir interface, which requires just a bit of manual work. So I think that 1.8 can be pronounced a success. Kudos to Marcelo.

In practise, the switch from 1.6 to 1.8 is happening in Linux distributions. 1.8 is the default in Fedora 29 and the 1.6 packages only appear as an explicit dependancy of something else, as far as I can see. Debian Buster has 1.8 behind the libupnp package, (it still has a libupnp6 package). Ubuntu Bionic and Cosmic have 1.8.2 package, Disco has 1.8.4 (they also have 1.6)

This goes to show that the current master/1.8 branch is in pass of becoming the main distribution/production branch, as decided by Linux distros and developpers.

As for changes which should not happen on this type of release branch, I have just created issues about 2 of them (the issue about the API change was mostly a heads-up, there is nothing that can be done about it, I think).

* The vdir API change which occurred between 1.8.2 and 1.8.3
* The runtime behaviour change linked to the reuseaddr configurability, which occurred between 1.8.0 and 1.8.1. Debian packages use --enable-reuseaddr

Once a version is packaged, Linux developers are stuck with supporting it for years.

Some other things which should not have gone so easily into a stable branch:

* My own change for allowing multiple root devices (not a bad change, just that it does not belong in a stable branch).
* The change in HTTP Status response format in 23f1d9346586ab815ce6ca9634e1e6d8064922ec . Except if there has been some behind the scene discussion, I don't see much reference to the HTTP standard or justification beyond 'compatibility with Samsung TVs'. Consequences for other devices ? Unknown (I actually think that the change is correct, I'm only pointing out the method).
* The move-around of threadutil code which happened between 1.8.0 and 1.8.1: at some point, this broke applications which included upnpdebug.h, because of a missing ThreadPool.h (to be once more clear: not saying that this is a bad change). There is no solution to this issue by the way: if the Linux distribution has this, the only possible approach is to change the system files.
* The change with Extra Headers did not get into master, but it did get into 1.6 (!). This caused crashes with the packaged 1.6.24 release, forcing the packagers to immediately patch or update to 1.6.25.

A few things I would like to do, which should not go into 1.8:

* Implementing support for multiple network interfaces.
* Debug log cleanup: there are some Debian patches for this, but I think that more work is needed, and I would not do this in a release branch.


Feature branches are nice, but they have to be merged somewhere for integration and wider testing.

To sum up my humble opinion:

* Branch 1.8 is becoming the production branch, as decided by Linux distributions and application authors.
* As a consequence, it should be considered stable and only receive bug and security fixes, after careful consideration. This would appear to support the creation of branch-1.8.x
* As another consequence, 1.6 can slowly go into obsolescence: security fixes only. It's working fine, no need to change it.
* We need a merge branch for risky changes and new features. master seems the natural place.

This is not a negative comment but an acknowledgement of the success of the 1.8 version.

**Comment by ukleinek** [comment]
Just to throw in my thoughts: I don't see a hard reason to branch of 1.8.x as of now. libupnp is not a fast-moving library. Yes, the API breaker was bad and should better have noticed before, but there was (TTBOMK) no big harm done. Looking at the list of things you consider too experimental for 1.8: I don't think they should not go into 1.8. With my Debian hat on I'd say: There are mechanisms that make it possible to handle breakage and given that there is some overhead for another main branch I'd say keep the dev model as is.

**Comment by ukleinek** [comment]
Looking at the situation for people using Debian, I think most of these problems doesn't affect these as the problems were noticed before the broken versions were uploaded to them. (Not sure about the upnpdebug.h problem you pointed out, though.)

**Comment by medoc92** [comment]
Actually, people running multiple instances of libupnp would have been affected by the reuseaddr change, but maybe 1.8.0 was never packaged ? Fedora 29 ships with the broken upnpdebug.h, and I think that Fedora has shipped with the crashing 1.6.24 for a time.

My point is that any development library needs a stable branch where point updates are no-brainers for packagers (and developers) because they can be confident that they only include bug fixes and well tested changes.

The question is: what branch is the stable one for libupnp  ? I thought that the time was ripe for 1.8, but then, maybe it's still 1.6 ? It would be nice if this could be clearly stated somewhere because it seems that there is a feeling around that 1.6 can go into obsolescence.

**Comment by mrjimenez** [comment]
I would like to talk about two things.

First: 

> The change with Extra Headers did not get into master, but it did get into 1.6 (!). 
> This caused crashes with the packaged 1.6.24 release, forcing the packagers to immediately
> patch or update to 1.6.25.

I don't have the time to check it now, but that is very odd. I have kept 1.8 as a 1.6 with hidden (encapsulated) data for a long time. I don't remember putting a feature in 1.6 that was not immediately included in 1.8. If you say so, I believe you, but we surely have a problem here, since 1.8 does not support a feature that 1.6 does support. Or maybe I got you wrong on this one?

Second, from a historical point of view, when I started 1.8, it was experimental. The truth is, if anyone had tested it, then I have not been told about it for a long time. Then people started including 1.8 in Debian. I have many times repeated that I would not support 1.8 because I had never had feedback on it. So eventually, mostly due to Debian packagers (thanks, guys!) we "officialized" 1.8 and started doing releases. From then on, the situation changed **completely**. Now people actually discuss 1.8 and 1.6 is almost forgotten.

That paragraph is to say that naming a new branch would probably have the effect of including it in a distribution, most likely Debian, even if it is totally experimental code. I don't expect people to have any king of "respect" for experimental releases, even because the code does not come "marked" with this tag. In the old days, odd versions were development, even versions were stable, but it seems to me that this is not really something universally accepted.

So, I am not against creating a new branch. But how do we balance the inclusion of new stuff with community testing without creating a branch that nobody ever uses, like it was with 1.8 for over 10 years? This situation is actually repeating itself right now. We have an extra-headers branch, that I consider it is a cool feature, and that no one is testing or giving feedback. At the same time, how do we include new stuff without terribly breaking stuff and make developers life miserable?

Just my two cents.
Best regards,
Marcelo.

**Comment by medoc92** [comment]
Responding in separate messages because this is getting complicated :)

About extra headers, 1.6 problems and why it never went into 1.8:

The crash  caused by somewhat hasty code in the extra headers features was introduced by this commit in july 2017: https://github.com/mrjimenez/pupnp/commit/9c2e8ec8a0291ebe81959009e2f78edbdb47ced5

The crash was fixed by this pull request: https://github.com/mrjimenez/pupnp/pull/59  (Dec 5 2017)
This one was also probably related: https://github.com/mrjimenez/pupnp/pull/62

The problem was in release 1.6.24 (19 Nov), and fixed by Joerg Krause (who beat me to it) on Dec 5. The fix release 1.6.25 was released on 10 Feb 2018

The current Extra Headers pull request (not a branch) was never merged into 1.8 because it's problematic there: https://github.com/mrjimenez/pupnp/pull/49

I forget stuff all the time too :) Long live the github logs !

**Comment by mrjimenez** [comment]
Hum, :)

Thanks for the forensics!

Well, I think we should move the extra-headers part of the discussion to the proper thread, so I will continue there.

Best regards,
Marcelo.

**Comment by medoc92** [comment]
> Second, from a historical point of view, when I started 1.8, it was experimental. The truth is, if anyone had tested it, then I have not been told about it for a long time. Then people started including 1.8 in Debian. I have many times repeated that I would not support 1.8 because I had never had feedback on it. So eventually, mostly due to Debian packagers (thanks, guys!) we "officialized" 1.8 and started doing releases. From then on, the situation changed **completely**. Now people actually discuss 1.8 and 1.6 is almost forgotten.

Agreed, this is what started the whole thing for me, I've had people asking me to port to 1.8 because 1.6 is going away. What changed the situation too was the brave soul (sorry too lazy to dig the name) who wrote umpteen macros to make the port easy.
It's while doing and testing the port that I realized that there were compatibility issues between successive 1.8 releases.

> That paragraph is to say that naming a new branch would probably have the effect of including it in a distribution, most likely Debian, even if it is totally experimental code. I don't expect people to have any king of "respect" for experimental releases, even because the code does not come "marked" with this tag. In the old days, odd versions were development, even versions were stable, but it seems to me that this is not really something universally accepted.

I think that many people still do something like even/odd _if there is a need for experimental releases_ . For example, Xapian, which is my personal reference of "how things should be done" does it this way (and the odd ones are probably only used internally). 

But  people also understand that the master branch is where new stuff gets merged. If you keep a master branch with no release tags on it, it will be used by people who need the latest stuff (so it will get tested), and it will quite probably not be packaged (in any case it will be very clear that you are not making any promise about it).

> So, I am not against creating a new branch. But how do we balance the inclusion of new stuff with community testing without creating a branch that nobody ever uses, like it was with 1.8 for over 10 years? This situation is actually repeating itself right now. We have an extra-headers branch, that I consider it is a cool feature, and that no one is testing or giving feedback. At the same time, how do we include new stuff without terribly breaking stuff and make developers life miserable?

Like I wrote above, Extra Headers is just a pull request for 1.8, it is active in 1.6, it did not get into 1.8 because of implementation issues. 
The main barrier to 1.8 adoption was that it was a serious pain to move 1.6 app code to 1.8. I don't think that this is likely to happen often.

If the new stuff goes to master, it will get used by the people who produce the changes because they are necessitated by their applications. While trying their new stuff, they will also _have_ to use the stuff from others. None of this will go to bother users, or application writers who don't follow closely, until there is a decision that it is stable enough and compatible enough to become the new production branch (1.10), at which point packagers will package (probably).

People expect breakage in master, they expect stability in named releases, especially patch (3rd number) ones.

After 1.8 is branched;
* branch 1.6 receives no changes except for serious security issues.
* branch 1.8 receives bug fixes and security fixes (no new features)
* All new features get merged into master only.

Hopefully, this should not cause more work for you. By the way, I don't think that this is a very original approach, but rather the most traditional way to do things, and, of course, just a proposition.
